File: Binary_Tree/437.Path_Sum_III/Python/437.Path_Sum_III.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def pathSum(self, root, targetSum):
        """
        :type root: Optional[TreeNode]
        :type targetSum: int
        :rtype: int
        """
        #======================================#
        # Recursion-based DFS traversal method #
        #======================================#
        if (not root): #Issue/Boundary-case handling
            return 0

        ############
        #Initialize
        ##### Record dictionary/hashmap #####
        record_dict = {0: 1}

        ##### Result summary counter #####
        res_cnter = [0]


        #####################################################################
        #Recursion-based DFS loop traversal with recorded dictionary/hashmap
        self.dfsTraversal(root, record_dict, 0, targetSum, res_cnter) #Recursion function call (i.e. whole root node)

        return res_cnter[0]


    def dfsTraversal(self, root, record_dict, record_prefix_sum, targetSum, res_cnter):
        """
        :type root: Optional[TreeNode]
        :type record_dict: dict
        :type record_prefix_sum: int
        :type targetSum: int
        :type res_cnter: List
        :rtype: None, void
        """
        #========================================#
        # Recursion-based DFS traversal function #
        #========================================#

        ############
        #Initialize
        ##### Current prefix-summary value #####
        curr_prefix_sum = (record_prefix_sum + root.val)

        ####################
        #Whole process/flow
        ##### Check if the current prefix-summary existed or not #####
        if ((curr_prefix_sum - targetSum) in record_dict):
            res_cnter[0] += record_dict[(curr_prefix_sum - targetSum)] #Update/Accumulate

        ##### Check if the current prefix-summary existed or not #####
        if (curr_prefix_sum not in record_dict):
            record_dict[curr_prefix_sum] = 1 #Update/Record
        else:
            record_dict[curr_prefix_sum] += 1 #Update/Record

        if (root.left):
            self.dfsTraversal(root.left, record_dict, curr_prefix_sum, targetSum, res_cnter) #Recursion function call (i.e. root left-half node)

        if (root.right):
            self.dfsTraversal(root.right, record_dict, curr_prefix_sum, targetSum, res_cnter) #Recursion function call (i.e. root right-half node)

        record_dict[curr_prefix_sum] -= 1 #Update/Record

File: Binary_Tree/437.Path_Sum_III/C++/437.Path_Sum_III.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int pathSum(TreeNode * root, int targetSum) 
    {
        //==================================//
        // Input type:                      //
        //   - TreeNode * (root tree-node)  //
        //   - int (target summary value)   //
        // Return type:                     //
        //   - int (result summary counter) //
        //==================================//
        if (! root) { return 0; } //Issue, Boundary-case handling

        /*Initialize*/
        ///// Record dictionary, hashmap /////
        unordered_map<long long, int> record_map;

        record_map[0] = 1;

        ///// Result summary counter /////
        int res_cnter = 0;


        /*Recursion-based DFS loop traversal*/
        dfsTraversal(root, record_map, 0, targetSum, res_cnter); //Recursion function call (i.e. whole root node)

        return res_cnter;
    }


    void dfsTraversal(TreeNode * root, unordered_map<long long, int> & record_map, long long record_prefix_sum, int & targetSum, int & res_cnter)
    {
        //=====================================================================//
        // Input type:                                                         //
        //   - TreeNode * (root tree-node)                                     //
        //   - unordered_map<long long, int> ref. (record dictionary, hashmap) //
        //   - int (record prefix-summary value)                               //
        //   - int ref. (target summary value)                                 //
        // Return type:                                                        //
        //   - void (no return)                                                //
        //=====================================================================//

        /*Initialize*/
        ///// Current prefix-summary value /////
        long long curr_prefix_sum = (record_prefix_sum + (root->val));

        /*Whole process, flow*/
        ///// Check if the current prefix-summary value existed or not /////
        if (record_map.find((curr_prefix_sum - targetSum)) != record_map.end())
        {
            res_cnter += record_map[(curr_prefix_sum - targetSum)]; //Keep updating, accumulating
        }

        ///// Check if the current prefix-summary value existed or not /////
        if (record_map.find(curr_prefix_sum) != record_map.end()) { record_map[curr_prefix_sum] += 1; } //Keep updating, recording
        else { record_map[curr_prefix_sum] = 1; } //Keep updating, recording

        if ((root->left)) { dfsTraversal((root->left), record_map, curr_prefix_sum, targetSum, res_cnter); } //Recursion function call (i.e. root left-half node)
        if ((root->right)) { dfsTraversal((root->right), record_map, curr_prefix_sum, targetSum, res_cnter); } //Recursion function call (i.e. root right-half node)

        record_map[curr_prefix_sum] -= 1; //Keep updating, recording
    }
};
